### PR TITLE
Introduce `NetworkConfig` on Android

### PR DESCRIPTION
**Pull request by rolandkakonyi**
## Description

This PR implements the APIs introduced in #449 for Android.

## Changes

Introduced `NetworkModule` on Android with all `NetworkConfig` support.

Note: it is expected that the iOS/tvOS CI builds fail, as we use an API that is introduced in the now in-progress release.

## Checklist
- [x] 🗒 `CHANGELOG` entry - already exists

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -99,6 +99,7 @@ dependencies {
     //noinspection GradleDynamicVersion
     implementation 'com.facebook.react:react-native:+' // From node_modules
     implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    implementation "androidx.concurrent:concurrent-futures-ktx:1.1.0"
```
Comment (rolandkakonyi) [inline android/build.gradle:102]: This is necessary for using `CallbackToFutureAdapter` in `NetworkModule.kt`

File: android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt [hunk 2]
```diff
@@ -785,6 +791,50 @@ fun ReadableMap.toLiveConfig(): LiveConfig = LiveConfig().apply {
...
+        getString("url") ?: return null,
+        getMap("headers")?.toMap(),
+        getString("body")?.toByteArrayFromBase64(),
+        getString("method") ?: return null,
+    )
+}
+
+private fun ByteArray.toBase64String(): String {
+    return Base64.encodeToString(this, Base64.NO_WRAP)
+}
```
Comment (123mpozzi) [inline android/src/main/java/com/bitmovin/player/reactnative/converter/JsonConverter.kt:805]: To add context here: I went for `Base64.NO_WRAP` as apparently `Base64.DEFAULT` adds a line break

**Review by 123mpozzi — COMMENTED** [review]
The code seems to be working 💪 
However I am no expert in Kotlin so I cannot comment the style 😅 : I see we use this `Completer` which is very similar to what we do in the iOS counterpart.

---

I am also leaving some additional notes to things we should look for:
- testing performance on a real device
- testing base64 encoding/decoding: to me it looks like the `body` fields here are bigger than in the iOS counterpart

**Comment by rolandkakonyi** [comment]
> * testing base64 encoding/decoding: to me it looks like the `body` fields here are bigger than in the iOS counterpart

@123mpozzi what do you mean here by `body` fields are bigger?
We can't really compare as iOS and Android loads different manifests (HLS vs DASH) AND additionally Android supports video and audio segments with the network API, while iOS doesn't.

**Review by matamegger — CHANGES_REQUESTED** [review]
Looks good overall 👍

File: android/src/main/java/com/bitmovin/player/reactnative/NetworkModule.kt [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            preprocessHttpResponseCompleters.remove(it)
+        }
+    }
+
+    private fun initConfigBlocks(nativeId: String, config: Any?) {
+        (config as? ReadableMap)?.let { json ->
+            initPreprocessHttpRequest(nativeId, networkConfigJson = json)
+            initPreprocessHttpResponse(nativeId, networkConfigJson = json)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply android/src/main/java/com/bitmovin/player/reactnative/NetworkModule.kt]: Good catch! Fixed in edf371a0

File: android/src/main/java/com/bitmovin/player/reactnative/BitmovinBaseModule.kt [hunk 4]
```diff
@@ -63,6 +68,12 @@ abstract class BitmovinBaseModule(
         nativeId: NativeId,
         sourceModule: SourceModule = this.sourceModule,
     ): Source = sourceModule.getSourceOrNull(nativeId) ?: throw IllegalArgumentException("Invalid SourceId $nativeId")
+
+    fun RejectPromiseOnExceptionBlock.getNetworkConfig(
+        nativeId: NativeId,
+        networkModule: NetworkModule = this.networkModule,
+    ): NetworkConfig = networkModule.getConfig(nativeId)
+        ?: throw IllegalArgumentException("Invalid NetworkId $nativeId")
```
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply android/src/main/java/com/bitmovin/player/reactnative/BitmovinBaseModule.kt]: Removed in 625594f2

File: android/src/main/java/com/bitmovin/player/reactnative/NetworkModule.kt [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            preprocessHttpResponseCompleters[responseId] = completer
+            context.catalystInstance.callFunction("Network-$nativeId", "onPreprocessHttpResponse", args as NativeArray)
+        }
+    }
+
+    @ReactMethod
+    fun setPreprocessedHttpResponse(responseId: String, response: ReadableMap) {
+        preprocessHttpResponseCompleters[responseId]?.set(response.toHttpResponse())
+        preprocessHttpResponseCompleters.remove(responseId)
+    }
```
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply android/src/main/java/com/bitmovin/player/reactnative/NetworkModule.kt:121]: I tried to use a single lock to wrap all writing of all of those maps but it affects the playback performance.

Do you maybe have a suggestion how could this be implemented in a performant way?

File: android/build.gradle (hunk 1 above)
Comment (rolandkakonyi) [reply android/build.gradle:103]: Added the missing dependency here: b0222168

File: android/src/main/java/com/bitmovin/player/reactnative/NetworkModule.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply android/src/main/java/com/bitmovin/player/reactnative/NetworkModule.kt:121]: As discussed offline, let's keep this as is for now.

File: android/src/main/java/com/bitmovin/player/reactnative/PlayerModule.kt [hunk 6]
```diff
@@ -69,6 +71,8 @@ class PlayerModule(context: ReactApplicationContext) : BitmovinBaseModule(contex
         val analyticsConfig = analyticsConfigJson?.toAnalyticsConfig()
         val defaultMetadata = analyticsConfigJson?.getMap("defaultMetadata")?.toAnalyticsDefaultMetadata()
 
+        networkNativeId?.let { networkModule.getConfig(it) }?.let { playerConfig.networkConfig = it }
```
[the remark this answers is not in the document]
Comment (rolandkakonyi) [reply android/src/main/java/com/bitmovin/player/reactnative/PlayerModule.kt]: Improved in a9685d8b

**Comment by 123mpozzi** [comment]
> > * testing base64 encoding/decoding: to me it looks like the `body` fields here are bigger than in the iOS counterpart
> 
> @123mpozzi what do you mean here by `body` fields are bigger? We can't really compare as iOS and Android loads different manifests (HLS vs DASH) AND additionally Android supports video and audio segments with the network API, while iOS doesn't.

I see, then the difference makes sense 👍

**Pull request merged**